OSSM-6260: Document migration from SMCP managed gateways   Part of Service Mesh 2.6 release (#78340)
openshift/openshift-docs: modules/ossm-migrating-from-smcp-defined-gateways-to-gateway-injection.adoc

// Module included in the following assemblies:
//
// * service_mesh/v2x/ossm-gateway-migration.adoc

:_mod-docs-content-type: CONCEPT
[id="ossm-migrating-from-smcp-defined-gateways-to-gateway-injection_{context}"]

> = Migrate from SMCP-Defined gateways to gateway injection

Author: @skopacz1 it's very common for customers to refer to the SMCP and know it means the ServiceMeshControlPleane resource. In this one instance, I would prefer to have the title be short.

> :context: ossm-migrating-from-smcp-defined-gateways-to-gateway-injection

Author: done

This procedure explains how to migrate with zero downtime from gateways defined in the `ServiceMeshControlPlane` resource to gateways that are managed using gateway injection. This migration is achieved by using the existing gateway Service resource to target a new gateway deployment that is created using gateway injection.

.Prerequisties

> * You are logged in to the {ocp-product-title} web console as `cluster-admin`.

Author: done

* The {SMProductName} Operator must be installed.

* The `ServiceMeshControlPlane` resource must be deployed and an ingress gateway exists in the configuration.

.Procedure

. Create a new ingress gateway that is configured to use gateway injection. 
+
[NOTE]
====
This procedure migrates away from the default ingress gateway deployment defined in the `ServiceMeshControlPlane` resource to gateway injection. The procedure may be modified to migrate from additional ingress gateways configured in the SMCP.    
====
+

> [source,yaml, subs="attributes,verbatim"]

Author: done

      istio: ingressgateway 
  template:
    metadata:
      annotations:
        inject.istio.io/templates: gateway
      labels: <2>
        app: istio-ingressgateway
        istio: ingressgateway
        sidecar.istio.io/inject: "true"
    spec:
      containers:
        - name: istio-proxy
          image: auto
      serviceAccountName: istio-ingressgateway
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: istio-ingressgateway
  namespace: istio-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: secret-reader
  namespace: istio-system
rules:
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "watch", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: istio-ingressgateway-secret-reader
  namespace: istio-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: secret-reader
subjects:
  - kind: ServiceAccount
    name: istio-ingressgateway
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy <3>
metadata:
  name: gatewayingress
  namespace: istio-system
spec:
  podSelector:
    matchLabels:
      istio: ingressgateway
  ingress:
    - {}
  policyTypes:
    - Ingress

> ---- 

Reviewer: It's out of my scope to check for technical accuracy, but is this all meant to be in one big YAML file? Or are these meant to be distinct YAML files that you create separately? If it's the latter, they need to be split out into separate files, and likely their own steps. Something like:

. Create a new ingress gateway that is configured to use gateway injection:

.. Create YAML file 1

.. Create YAML file 2

etc

Ignore this comment if that's not the case, and let me know if you have questions about what I mean here

Author: This is one big YAML file. I considered linking to a repo that contains the file, but we decided to put all the code in the topic.

<1> The gateway injection deployment and all supporting resources should be deployed in the same namespace as the SMCP-defined gateway. 
<2> Ensure that the labels specified in the pod template include all of the label selectors specified in the Service associated with the existing SMCP-defined gateway.
<3> Grant access to the new gateway from outside the cluster. This access is required whenever the `spec.security.manageNetworkPolicy` of the `ServiceMeshControlPlane` resource is set to `true`, which is the default setting.

> . Verify that the new gateway deployment is successfully handling requests. 
> +
> Ensure that the new gateway deployment is successfully handling requests. If access logging was configured in the `ServiceMeshControlPlane` resource view the access logs of the new gateway deployment to confirm the behavior.

Reviewer: The first and second sentence of this step seem to be saying the same thing, I think one of them should be deleted:
```suggestion
. Verify that the new gateway deployment is successfully handling requests. 
+
If access logging was configured in the `ServiceMeshControlPlane` resource view the access logs of the new gateway deployment to confirm the behavior.
```

Author: done

Author: lol yup. done. thank you for catching this.

> Ensure that the new gateway deployment is successfully handling requests. If access logging was configured in the `ServiceMeshControlPlane` resource view the access logs of the new gateway deployment to confirm the behavior.

Reviewer: Suggestion to add a comma here for clarity:
```suggestion
Ensure that the new gateway deployment is successfully handling requests. If access logging was configured in the `ServiceMeshControlPlane` resource, view the access logs of the new gateway deployment to confirm the behavior.
```

Author: done

> . Scale down the old deployment and scale up the new deployment. 
> +
> You may gradually shift traffic from the old gateway deployment to the new gateway deployment by running the `oc scale` command at the command line:
> +
> [source,terminal]
> ----
> $ oc scale -n istio-system deployment/<gateway_deployment> --replicas <new_number_of_replicas>
> ----
> +
> As you scale, increase the number of replicas for the new gateway deployment and decrease the number of replicas for the old gateway deployment. Continue scaling until the new gateway deployment handles all traffic to the gateway Service. 

Reviewer: I'm a little confused by the nature of this step. Is this meant to be optional? I don't think it is, but either way "You may" should probably be removed from the second sentence, since you're instructing users to run a command. If I'm interpreting this step right, where you repeat this command while increasing the replica values "until the new gateway deployment handles all traffic to the gateway Service", then I would suggest writing the step like this:

```suggestion
. Scale down the old deployment and scale up the new deployment. 
+
Gradually shift traffic from the old gateway deployment to the new gateway deployment by performing the following steps:

.. Increase the amount of replicas for the new gateway deployment by running the following command:
+
[source,terminal]
----
$ oc scale -n istio-system deployment/<new_gateway_deployment> --replicas <new_number_of_replicas>
----

.. Decrease the amount of replicas for the old gateway deployment by running the following command:
+
[source,terminal]
----
$ oc scale -n istio-system deployment/<old_gateway_deployment> --replicas <new_number_of_replicas>
----

 .. Run the previous two commands, each time increasing the number of replicas for the new gateway deployment and decreasing the number of replicas for the old gateway deployment, until the new gateway deployment handles all traffic to the gateway Service. 
```

This is just one way to rework this step, but the point is that I think breaking this down into specific, granular steps, where each step only has one explicit action for users to follow, would make this clearer for users to follow IMO. There are probably other ways to rework this, so let me know if you'd like to discuss this more!

Author: done

> . Remove the `app.kubernetes.io/managed-by` label from the gateway Service resource.
> +
> Removing the label prevents the Service from being deleted when the gateway is disabled in the `ServiceMeshControlPlane` resource. Run the following command at the CLI:
> +
> [source,terminal]
> ----
> $ oc label service -n istio-system istio-ingressgateway app.kubernetes.io/managed-by-
> ----

Reviewer: Up to you, but suggestion to move the context of this command to the end of this step:
```suggestion
. Remove the `app.kubernetes.io/managed-by` label from the gateway Service resource by running the following command:
+
[source,terminal]
----
$ oc label service -n istio-system istio-ingressgateway app.kubernetes.io/managed-by-
----
+
Removing the label prevents the Service from being deleted when the gateway is disabled in the `ServiceMeshControlPlane` resource.
```

Author: done

> . Remove the `ownerReferences` object from the gateway Service resource.
> +
> Removing this object prevents the Service from being garbage collected when the `ServiceMeshControlPlane` resource is deleted.
> +
> [source,terminal]
> ----
> $ oc patch service -n istio-system istio-ingressgateway --type='json' -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
> ----

Reviewer: Similar suggestion as above. Again, up to you:
```suggestion
. Remove the `ownerReferences` object from the gateway Service resource by running the following command:
+
[source,terminal]
----
$ oc patch service -n istio-system istio-ingressgateway --type='json' -p='[{"op": "remove", "path": "/metadata/ownerReferences"}]'
----
+
Removing this object prevents the Service from being garbage collected when the `ServiceMeshControlPlane` resource is deleted.
```

Reviewer: If you don't take that suggestion, then you must at least add a mention of running a command somewhere in this step

Author: done

. Disable the old gateway deployment that was managed by the `ServiceMeshControlPlane` resource by running the following command at the CLI:
+
[source,terminal]
----
$ oc patch smcp -n istio-system <smcp_name> --type='json' -p='[{"op": "replace", "path": "/spec/gateways/ingress/enabled", "value": false}]'
----
+

> When the old ingress gateway Service resource is disabled it is not deleted. You may save this Service resource to a file and manage it alongside the new gateway injection resources.

Reviewer: Just checking - is this info something you think could be wrapped in a NOTE?

Author: done

:_mod-docs-content-type: CONCEPT
[id="ossm-migrating-from-smcp-defined-gateways-to-gateway-injection_{context}"]
= Migrate from SMCP-Defined gateways to gateway injection

This procedure explains how to migrate with zero downtime from gateways defined in the `ServiceMeshControlPlane` resource to gateways that are managed using gateway injection. This migration is achieved by using the existing gateway Service resource to target a new gateway deployment that is created using gateway injection.

.Prerequisties

* You are logged in to the {product-title} web console as `cluster-admin`.

* The {SMProductName} Operator must be installed.

* The `ServiceMeshControlPlane` resource must be deployed and an ingress gateway exists in the configuration.

.Procedure

. Create a new ingress gateway that is configured to use gateway injection. 
+
[NOTE]
====
This procedure migrates away from the default ingress gateway deployment defined in the `ServiceMeshControlPlane` resource to gateway injection. The procedure may be modified to migrate from additional ingress gateways configured in the SMCP.    
====
+
.Example ingress gateway resource with gateway injection

> +

Reviewer 2: Not needed between a block title and the block options:
```suggestion
```

Author: done

spec:
  podSelector:
    matchLabels:
      istio: ingressgateway
  ingress:
    - {}
  policyTypes:
    - Ingress
---- 
<1> The gateway injection deployment and all supporting resources should be deployed in the same namespace as the SMCP-defined gateway. 
<2> Ensure that the labels specified in the pod template include all of the label selectors specified in the Service associated with the existing SMCP-defined gateway.
<3> Grant access to the new gateway from outside the cluster. This access is required whenever the `spec.security.manageNetworkPolicy` of the `ServiceMeshControlPlane` resource is set to `true`, which is the default setting.

. Verify that the new gateway deployment is successfully handling requests. 
+
If access logging was configured in the `ServiceMeshControlPlane` resource, view the access logs of the new gateway deployment to confirm the behavior.

. Scale down the old deployment and scale up the new deployment. 
+
Gradually shift traffic from the old gateway deployment to the new gateway deployment by performing the following steps:

.. Increase the number of replicas for the new gateway deployment by running the following command:
+
[source,terminal]
----

> $ oc scale -n istio-system deployment/<gateway_deployment> --replicas <new_number_of_replicas>

Reviewer: Unless this is intentional, this value still needs to have "new" added to it:
```suggestion
$ oc scale -n istio-system deployment/<new_gateway_deployment> --replicas <new_number_of_replicas>
```

Author: done

----

> +

Reviewer 2: Not needed between list items, typically:
```suggestion

```

Author: done

.. Decrease the number of replicas for the old gateway deployment by running the following command:
+
[source,terminal]
----
$ oc scale -n istio-system deployment/<old_gateway_deployment> --replicas <new_number_of_replicas>
----

.. Repeat running the previous two commands. Each time, increase the number of replicas for the new gateway deployment and decrease the number of replicas for the old gateway deployment. Continue repeating until the new gateway deployment handles all traffic to the gateway Service. 

. Remove the `app.kubernetes.io/managed-by` label from the gateway Service resource by running the following command:
+
[source,terminal]
----

> $ oc label service -n istio-system istio-ingressgateway app.kubernetes.io/managed-by-

Reviewer: technical accuracy is out of my scope, but is the hyphen at the end here intentional?
```suggestion
$ oc label service -n istio-system istio-ingressgateway app.kubernetes.io/managed-by
```

Author: @skopacz1 Yes, that dash at the end is the correct code. I had the same conversation with the developer lol.
